django/db/models/query.py

```diff
@@ -1606,8 +1609,11 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
                 setattr(obj, to_attr, vals)
                 obj._prefetched_objects_cache[cache_name] = vals
             else:
-                # Cache in the QuerySet.all().
-                qs = getattr(obj, to_attr).all()
+                manager = getattr(obj, to_attr)
+                if leaf and lookup.queryset is not None:
+                    qs = manager._filter_queryset(lookup.queryset)
+                else:
+                    qs = manager._filter_queryset(rel_qs)
```

Comment:
By using `rel_qs` we can simplify to: `qs = getattr(obj, to_attr)._filter_queryset(rel_qs)` and do away with the `if/else`. What do you think?